kuzzleio/kuzzle
# Allow to check user rights when using EmbeddedSDK.as impersonation

## What does this PR do ?

Long story short. :fast_forward: 
Updates the `app.sdk.as` system to be less expensive and also implements a **conditional** security check _(user rights)_ before executing any action at the funnel level.

---

Current impersonation using Kuzzle's internal **EmbeddedSDK** (`app.sdk.as()`) is actually quite expensive, it would instantiate a whole new SDK with all its bells and whistles. This PR changes this by implementing a lazily dynamic meta-constructed sdk wrapper called the **ImpersonatedSDK** that is considerably smaller, grows depending on the impersonated user's usage and most importantly it works by using the already instantiated base **sdk**.

### _To Dos_
- [x] **ImpersonatedSDK** Tests
- [x] Update **EmbeddedSDK** Tests
- [x] Update **funnelProtocol** Tests
- [x] Functional **Plugin Context** Create Document As

### How should this PR be reviewed?

- :warning: Please invest particular attention into finding any breaking differences between the **EmbeddedSDK** and our new **ImpersonatedSDK**. _(Atm, the ImpersonatedSDK only proxies the **controllers** + **query** actions, would that really suffice?)._ :warning: 
- Taking a look at the modifications of the **sdk/funnelProtocol**.
- Check out the below test example to see if any other weird **case** comes to your mind.
- Nvm **sonarcloud** :man_facepalming: 

### How should this be manually tested?

1. With kuzzle services running, `npm run dev test.ts`
```ts
import { Backend } from './lib/core/application/backend'

const app = new Backend('playground')

function assert(response, wantedId) {
    if (response._source._kuzzle_info.author !== wantedId)
        throw new Error(`Failed, ${response._source._kuzzle_info.author}!=${wantedId}`)
}

app.start()
.then(async () => {
    const proxyInstance = app.sdk.as({ _id: 'leo' }, { checkRights: true })

    assert(await app.sdk.document.create('test', 'test', { authorShouldBe: 'nobody' }), null)
    assert(await proxyInstance.document.create('test', 'test', { authorShouldBe: 'leo' }), 'leo')
    assert(await app.sdk.document.create('test', 'test', { authorShouldBe: 'nobody' }), null)
    assert(await app.sdk.as({ _id: 'alyx' }).document.create('test', 'test', { authorShouldBe: 'alyx' }), 'alyx')
    assert(await app.sdk.document.create('test', 'test', { authorShouldBe: 'nobody' }), null)
    assert(await app.sdk.as({ _id: 'leo' }).document.create('test', 'test', { authorShouldBe: 'leo' }), 'leo')
    assert(await app.sdk.document.create('test', 'test', { authorShouldBe: 'nobody' }), null)

    // This doesn't work, you can't use '.as' with an already impersonated sdk instance.
    // await proxyInstance.as({ _id: 'alyx' }).document.create('test', 'test', { authorShouldBe: 'alyx' });

    // Should these be also made to work?
    // console.log(proxyInstance.protocol, proxyInstance.sdkVersion, proxyInstance.sdkName);
    // console.log(app.sdk.protocol, app.sdk.sdkVersion, app.sdk.sdkName);

})
.catch(console.error)
```

### Other changes

- Updated the **FunnelProtocol** to be able to handle performant user Impersonation on demand. _(With backwards compatibility)_
- Added **app** _(which is a Backend instance)_ to the **global object**. :flags: 
- Moved and renamed native Controller files. _(api/controllers/)_ and their tests!
- Had to update the UTs for the newly dissected **BackendPlugin**, Backend is a singleton now?

### Boyscout

- Removes a UT step duplicate (authController.test)
- Adds a new assert **assertionError** when trying to re-define the **global.app** instance (:warning: This makes the Backend class behave like a single instance only! (hard singleton)

Closes #1686
